sdk/core/azure-core/inc/azure/core/context.hpp

```diff
@@ -245,10 +336,24 @@ namespace Azure { namespace Core {
       }
     }
 
-    /**
-     * @brief The application context (root).
+    /** @brief Creates a new root Context. This context is not a child of any other context, and
+     * thus can be independently cancelled.
+     *
+     * @return New root context.
+     *
+     */
+    static Context CreateNewRoot() { return Context{std::make_shared<ContextSharedState>()}; }
+
+    /** @brief The `ApplicationContext` is the root of all Context objects.
+     *
+     * @note Cancelling the `ApplicationContext` will cancel ALL operations currently active in the
+     * process, unless they were created with Azure::Core::Context::CreateNewRoot().
+     *
+     * @note If the `ApplicationContext` is cancelled, all subsequent operations will be cancelled.
```

Comment:
Would it make more sense to customers to have `Reset()` method which clears the cancellation? Thus you can resume using the ApplicationContext rather than forcing the users to use a new rooted context?

Reply:
I chatted with Jeff about this, he and I strongly feel that you cannot undo a cancellation - that's because you have no way of knowing if all the operations you're cancelling have checked to see if their context is cancelled.

The only way to "undo" a cancellation is to create a new rooted context. 

If the "Reset" API was a "Create new rooted context on this Context", it would work, but we still need the "CreateNewRoot()" API to create a new Context that isn't rooted in ApplicationContext. So then we have to answer the question of "why are there two ways of creating a new rooted context?"

Reply:
I disagree that knowing if it was cancelled is a requirement of the design. The guarantees on cancelling are weak even without this change. As a developer I don't know where I am in the call stack when cancelling is checked, Only by reading code can you infer that cancelling does not prevent an operation from going to server and waiting for it to complete. In fact it may have gone, succeeded, and be sending you back the response result only to be interrupted with a cancelled message. There is no easy programmatic way to know the operation did successfully send and complete on the server.

Side note, I wonder if the design should only be to cancel up till we send or when we are in a loop polling. Otherwise cancellation is allowed to return results back to the developer.

Perhaps ApplicationContext is immutable and if you want the ability to cancel you must first create a child from it? Or maybe we just don't have ApplicationContext connected to anything and we document and deprecate it. IMHO it causes more problems than it solves.

Reply:
Today, ApplicationContext is immutable, to cancel it, you create a copy of it (which creates a new mutable reference to the ApplicationContext which can be cancelled). That design is unintuitive IMHO. In main, the behavior is: You can't cancel ApplicationContext (because it's immutable), so you cancel ApplicationContext by creating a *copy* of ApplicationContext, and cancel the *copy*, which has the effect of cancelling ApplicationContext?

In other words, to cancel ApplicationContext, you have to write:

```cpp
auto mutableContext = Azure::Core::Context::ApplicationContext;
mutableContext.Cancel()
```

which cancels `ApplicationContext`. 

This is literally the behavior that drove me into this entire "let's take a hard look at Context" thread - it was so unintuitive that I couldn't figure out how to explain that behavior to a customer.

Reply:
One more thought w.r.t. cancellation (a separate comment from the previous one because it's a different thought).

It has always been possible that a cancelled operation will return successfully to the caller. That's inherent in all cancellation models. But it does imply a new guideline: ServiceClients SHOULDNOT check for cancellation after the operation has completed to the point where it is reasonable to return values to the client.

And I have *absolutely* no problem with deprecating ApplicationContext - that's literally why I started this discussion with "What's the point of ApplicationContext" - I was thinking of deprecating it. 

IMHO, there are two possible solutions: We make ApplicationContext meaningful as a single point of cancellation, or we deprecate it. If we want to make ApplicationContext meaningful, it means we need to figure out a story that involves undoing a Cancel (because otherwise ApplicationContext is a single shot entity).

So maybe the end result of all of this is to simply deprecate ApplicationContext, and let developers choose to implement a process (or thread) wide Context?

Reply:
> Today, ApplicationContext is immutable, to cancel it, you create a copy of it (which creates a new mutable reference to the ApplicationContext which can be cancelled). That design is unintuitive IMHO. In main, the behavior is: You can't cancel ApplicationContext (because it's immutable), so you cancel ApplicationContext by creating a _copy_ of ApplicationContext, and cancel the _copy_, which has the effect of cancelling ApplicationContext?
> 
> In other words, to cancel ApplicationContext, you have to write:
> 
> ```c++
> auto mutableContext = Azure::Core::Context::ApplicationContext;
> mutableContext.Cancel()
> ```
> 
> which cancels `ApplicationContext`.
> 
> This is literally the behavior that drove me into this entire "let's take a hard look at Context" thread - it was so unintuitive that I couldn't figure out how to explain that behavior to a customer.

Complete agreeement with you on how unintuitive this is.